RavenDB-21764 Move identities to react

M4xymm opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-21764/Move-identities-to-react

### Additional description

Moved identities view from knockout to react.js with slightly updated layout.


### Type of change

- [ ] Bug fix
- [ ] Regression bug fix
- [x] Optimization
- [ ] New feature

### How risky is the change?

- [x] Low 
- [ ] Moderate 
- [ ] High
- [ ] Not relevant

### Backward compatibility

- [x] Non breaking change
- [ ] Ensured. Please explain how has it been implemented?
- [ ] Breaking change
- [ ] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [x] No

### Documentation update

- [ ] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [x] No documentation update is needed 

### Testing by Contributor

- [x] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [x] It has been verified by manual testing

### Testing by RavenDB QA team

- [ ] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [x] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [x] No

### UI work

- [ ] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [x] No UI work is needed

M4xymm committed:
RavenDB-21764 feat (move identities to react): added utils function like service and mocks

RavenDB-21764 feat (move identities to react): added validation

RavenDB-21764 feat (move identities to react): added utils function for mocks

RavenDB-21764 feat (move identities to react): added main component DocumentIdentities.tsx along with modal and AboutView

RavenDB-21764 feat (move identities to react): added storybook and tests

RavenDB-21764 feat (move identities to react): eslint fixes

RavenDB-21764 feat (move identities to react): studio conventions fixes

RavenDB-21764 feat (move identities to react): fixed RichAlert.tsx

RavenDB-21764 feat (move identities to react): CR fixes

RavenDB-21764 feat (move identities to react): Removed old knockout views

RavenDB-21764 feat (move identities to react): Fixed table width, removed formField array

RavenDB-21764 feat (move identities to react): clean up imports

RavenDB-21764 feat (move identities to react): CR fixes (fixed table, removed unnecessary props, fixed edit refetching, renamed consts to maintain consistency in base)

RavenDB-21764 feat (move identities to react): added types to useForm, small refactor for useForm to make it more clear

kalczur commented on src/Raven.Studio/typescript/components/common/shell/accessManagerSliceSelectors.ts:
```diff
@@ -169,6 +194,7 @@ export const accessManagerSelectors = {
     isClusterAdminOrClusterNode: selectIsClusterAdminOrClusterNode,
     getHasDatabaseAdminAccess: selectGetHasDatabaseAccessAdmin,
     getHasDatabaseWriteAccess: selectGetHasDatabaseAccessWrite,
+    getHasDatabaseReadAccess: selectGetHasDatabaseAccessRead,
```
It's not used anywhere. I don't think we need that, because all users have at least read access.

kalczur commented on src/Raven.Studio/typescript/components/pages/database/documents/identities/DocumentIdentities.stories.tsx:
```diff
@@ -0,0 +1,41 @@
+        accessManager.with_databaseAccess({
+            [db.name]: args.databaseAccess,
+        });
+        license.with_License();
```
license is probably not needed in this view

kalczur commented on src/Raven.Studio/typescript/components/pages/database/documents/identities/DocumentIdentities.tsx:
```diff
@@ -0,0 +1,122 @@
+
+    const { identitiesColumns } = useDocumentIdentitiesColumns(width);
+    const databaseAccessWrite = useAppSelector(accessManagerSelectors.getHasDatabaseWriteAccess)();
+    const [isOpen, setIsOpen] = React.useState(false);
```
Let's use `useBoolean` hook. You will need only `value` and `toggle` from it.

kalczur commented on src/Raven.Studio/typescript/components/pages/database/documents/identities/DocumentIdentities.tsx:
```diff
@@ -0,0 +1,122 @@
+    const databaseName = useAppSelector(databaseSelectors.activeDatabaseName);
+
+    const { identitiesColumns } = useDocumentIdentitiesColumns(width);
+    const databaseAccessWrite = useAppSelector(accessManagerSelectors.getHasDatabaseWriteAccess)();
```
For booleans please add 'has' prefix in this case. In general it's easier to read when you have prefix such as 'is', 'has' or 'can'.

kalczur commented on src/Raven.Studio/typescript/components/pages/database/documents/identities/DocumentIdentities.tsx:
```diff
@@ -0,0 +1,122 @@
+}
+
+function DocumentIdentitiesWithSize({ width }: DocumentIdentitiesWithSizeProps) {
+    const databaseName = useAppSelector(databaseSelectors.activeDatabaseName);
```
You are getting it here only to pass it to `useGetIdentities`. Let's move it to this hook.

kalczur commented on src/Raven.Studio/typescript/components/pages/database/documents/identities/DocumentIdentities.tsx:
```diff
@@ -0,0 +1,122 @@
+        },
+    });
+
+    const debouncedSearch = debounce((e: ChangeEvent<HTMLInputElement>) => setGlobalFilter(e.target.value), 300);
```
It should be wrapped in useMemo, because every re-render will recreate this and it will cause problems with debounce.

kalczur commented on src/Raven.Studio/typescript/components/pages/database/documents/identities/DocumentIdentities.tsx:
```diff
@@ -0,0 +1,122 @@
+                        </Col>
+                        <Col md={12} lg={4} xl={3}>
+                            <Input onChange={debouncedSearch} placeholder="Filter prefix" />
+                        </Col>
```
Now we have filter for each column build in VirtualGrid. We can get rid of it.

![image](https://github.com/user-attachments/assets/74132f5e-7ac2-473a-94b0-5c54a87588ba)

kalczur commented on src/Raven.Studio/typescript/components/pages/database/documents/identities/DocumentIdentities.tsx:
```diff
@@ -0,0 +1,122 @@
+
+    return (
+        <>
+            <div className="content-margin">
```
Here we are already in "content-padding", so "content-margin" can be skipped.

kalczur commented on src/Raven.Studio/typescript/components/pages/database/documents/identities/DocumentIdentities.tsx:
```diff
@@ -0,0 +1,122 @@
+                                table={identitiesTable}
+                                className="mt-3"
+                                isLoading={loading}
+                                heightInPx={800}
```
You can get height from <SizeGetter /> the same way as width

kalczur commented on src/Raven.Studio/typescript/components/pages/database/documents/identities/DocumentIdentities.tsx:
```diff
@@ -0,0 +1,122 @@
+
+    const toggleModal = (value: boolean) => setIsOpen(value);
+
+    const { loading, identities } = useGetIdentities(databaseName);
```
```suggestion
    const { isLoading, identities } = useGetIdentities(databaseName);
```

kalczur commented on src/Raven.Studio/typescript/components/pages/database/documents/identities/DocumentIdentities.tsx:
```diff
@@ -0,0 +1,122 @@
+                                disabled={!databaseAccessWrite}
+                                className="add-new-identity-btn py-2"
+                            >
+                                <Icon icon="plus" className="mr-2" />
```
Icon already have margin by default.
```suggestion
                                <Icon icon="plus" />
```

kalczur commented on src/Raven.Studio/typescript/components/pages/database/documents/identities/DocumentIdentities.tsx:
```diff
@@ -0,0 +1,122 @@
+                    }))
+                    .filter(Boolean)
+            ),
+    });
```
`useAsync` already stores result so you don't need to do it. It is in `asyncGetIdentities.result`.
```suggestion
    const asyncGetIdentities = useAsync(async () => {
        const result = await databasesService.getIdentities(database);
        
        return Object.keys(result)
                    .map((identity) => ({
                        prefix: identity,
                        value: result[identity],
                    }));
    }, []);
```

kalczur commented on src/Raven.Studio/typescript/components/pages/database/documents/identities/DocumentIdentities.tsx:
```diff
@@ -0,0 +1,122 @@
+                        prefix: identity,
+                        value: result[identity],
+                    }))
+                    .filter(Boolean)
```
After mapping there is no way to have nullish value here. Calling `filter` can be skipped

kalczur commented on src/Raven.Studio/typescript/components/pages/database/documents/identities/DocumentIdentitiesAddModal.tsx:
```diff
@@ -0,0 +1,157 @@
+        resolver: addIdentitiesYupResolver,
+        defaultValues,
+    });
+    const [isLoading, setIsLoading] = useState(false);
```
You can get `formState.isSubmitting` from `form`

kalczur commented on src/Raven.Studio/typescript/components/pages/database/documents/identities/DocumentIdentitiesAddModal.tsx:
```diff
@@ -0,0 +1,157 @@
+    const isEditing = !!defaultValues;
+
+    const formValues = useWatch({ control: form.control });
+    const onSubmit: SubmitHandler<AddIdentitiesFormData> = useCallback(
```
Let's return `tryHandleSubmit`

kalczur commented on src/Raven.Studio/typescript/components/pages/database/documents/identities/DocumentIdentitiesAddModal.tsx:
```diff
@@ -0,0 +1,157 @@
+                    <div className="w-100 d-flex flex-column gap-4 mb-4">
+                        <DocumentIdentitiesAddModalForm isEditing={isEditing} control={form.control} />
+                    </div>
+                    {formValues?.value && formValues.prefix && <InformationBadge />}
```
For value or prefix equal `0` you will have a bug

kalczur commented on src/Raven.Studio/typescript/components/pages/database/documents/identities/DocumentIdentitiesAddModal.tsx:
```diff
@@ -0,0 +1,157 @@
+                <ModalFooter>
+                    <Button
+                        className="border-0 btn-dark text-muted"
+                        onClick={() => props.toggleModal(false)}
```
```suggestion
                        onClick={props.toggleModal}
```

kalczur commented on src/Raven.Studio/typescript/components/pages/database/documents/identities/DocumentIdentitiesAddModal.tsx:
```diff
@@ -0,0 +1,157 @@
+
+function InformationBadge() {
+    return (
+        <div
+            className={classNames(
+                "bg-faded-info rounded-2 mt-3 align-items-center px-2 py-1 d-flex me-2 align-self-start"
+            )}
+        >
```
Use RichAlert component

kalczur commented:
Please remove old knockout views

kalczur commented on src/Raven.Studio/typescript/components/pages/database/documents/identities/DocumentIdentities.tsx:
```diff
@@ -0,0 +1,127 @@
+}
+
+export default function DocumentIdentities() {
+    const { value: isOpen, toggle: setIsOpen } = useBoolean(false);
```
```suggestion
    const { value: isOpen, toggle: toggleIsOpen } = useBoolean(false);
```

kalczur commented on src/Raven.Studio/typescript/components/pages/database/documents/identities/DocumentIdentitiesAddModal.tsx:
```diff
@@ -0,0 +1,187 @@
+
+    const formValues = useWatch({ control: form.control });
+
+    useEffect(() => {
+        const prefixExists = identities?.some(
+            (identity) => identity.prefix?.toLocaleLowerCase().slice(0, -1) === formValues.prefix?.toLocaleLowerCase()
+        );
+
+        if (prefixExists) {
+            form.setError("prefix", { type: "custom", message: "A prefix with the same name exists" });
+        } else {
+            form.clearErrors("prefix");
+        }
+    }, [formValues, identities, form]);
```
Move this logic to yup schema

kalczur commented on src/Raven.Studio/typescript/components/pages/database/documents/identities/DocumentIdentitiesAddModal.tsx:
```diff
@@ -0,0 +1,187 @@
+        return tryHandleSubmit(async () => {
+            await databasesService.seedIdentity(databaseName, prefix, value);
+            form.reset();
+            if (isFunction(refetch)) refetch();
```
Let's maintain consistency in the codebase and add braces
```suggestion
            if (isFunction(refetch)) {
                refetch();
            }
```

kalczur replied:
Does this mean that the data in the table will not be refreshed after editing?

kalczur commented on src/Raven.Studio/typescript/components/pages/database/documents/identities/DocumentIdentitiesAddModal.tsx:
```diff
@@ -0,0 +1,187 @@
+
+    return (
+        <Modal
+            contentClassName={classNames("modal-border bulge-primary", contentClassName)}
+            wrapClassName={classNames("bs5", wrapClassName)}
```
contentClassName and wrapClassName are leftovers?

kalczur commented on src/Raven.Studio/typescript/components/pages/database/documents/identities/DocumentIdentitiesAddModal.tsx:
```diff
@@ -0,0 +1,187 @@
+function DocumentIdentitiesAddModalForm({ control, isEditing }: DocumentIdentitiesAddModalFormProps) {
+    return (
+        <>
+            <InputGroup className="vstack my-1">
```
Use `<FormGroup>`instead

kalczur commented on src/Raven.Studio/typescript/components/pages/database/documents/identities/DocumentIdentities.tsx:
```diff
@@ -0,0 +1,127 @@
+    return (
+        <div className="content-padding">
+            <Col md={12} lg={7} className="h-50">
+                <SizeGetter
+                    render={(props) => (
+                        <DocumentIdentitiesWithSize
+                            setIsOpen={setIsOpen}
+                            identities={identities}
+                            reload={reload}
+                            status={status}
+                            isLoading={isLoading}
+                            {...props}
+                        />
+                    )}
+                />
```
This probably causes the scroll to appear because the table is too large

kalczur reviewed (approved):
It looks very nice, good job!

kalczur commented on src/Raven.Studio/typescript/components/pages/database/documents/identities/DocumentIdentitiesModal.tsx:
```diff
@@ -0,0 +1,174 @@
+                {
+                    identities,
+                    isEditing,
+                },
```
Optional: you can add `satisfies DocumentIdentitiesPrefixTestContext`

Danielle9897 committed:
RDoc-3139 Create About Text for Identities View

ml054 commented on src/Raven.Studio/typescript/components/pages/database/documents/identities/DocumentIdentities.spec.tsx:
```diff
@@ -0,0 +1,52 @@
+
+describe("DocumentIdentities", () => {
+    beforeAll(() => {
+        Object.defineProperty(HTMLElement.prototype, "scrollWidth", {
```
let's move to setup_jest.js

ml054 commented on src/Raven.Studio/typescript/components/pages/database/documents/identities/DocumentIdentities.spec.tsx:
```diff
@@ -0,0 +1,52 @@
+        });
+    });
+
+    it("should render component with sample data and not be able to click btn", () => {
```
rename to have more meaningful names

ml054 commented on src/Raven.Studio/typescript/components/pages/database/documents/identities/DocumentIdentities.tsx:
```diff
@@ -0,0 +1,153 @@
+import { LoadError } from "components/common/LoadError";
+import { AddIdentitiesFormData } from "components/pages/database/documents/identities/DocumentIdentitiesValidation";
+
+interface DocumentIdentitiesWithSizeProps {
```
move down

ml054 commented on src/Raven.Studio/typescript/components/pages/database/documents/identities/DocumentIdentitiesModal.tsx:
```diff
@@ -0,0 +1,171 @@
+        const clientConfiguration = manageServerService.getClientConfiguration(databaseName);
+        const globalClientConfiguration = manageServerService.getGlobalClientConfiguration();
+
+        const [clientConfig, globalConfig] = await Promise.allSettled([clientConfiguration, globalClientConfiguration]);
```
handle 500 http code.

ml054 commented on src/Raven.Studio/typescript/components/pages/database/documents/identities/DocumentIdentitiesValidation.ts:
```diff
@@ -0,0 +1,32 @@
+    identities: Identity[];
+}
+
+const normalizeString = (value: string) => {
```
move to common ?

ml054 commented on src/Raven.Studio/typescript/components/pages/database/documents/identities/useDocumentIdentitiesColumns.tsx:
```diff
@@ -0,0 +1,64 @@
+            header: "Document ID Prefix",
+            accessorKey: "prefix",
+            cell: CellValueWrapper,
+            size: getSize(!databaseAccessWrite ? 50 : 46),
```
databaseAccessWrite ? 46: 50

ml054 commented on src/Raven.Studio/typescript/components/pages/database/documents/identities/useDocumentIdentitiesColumns.tsx:
```diff
@@ -0,0 +1,64 @@
+            header: "Latest value",
+            accessorKey: "value",
+            cell: CellValueWrapper,
+            size: getSize(!databaseAccessWrite ? 50 : 46),
```
same here

ml054 commented on src/Raven.Studio/typescript/components/pages/database/documents/identities/useDocumentIdentitiesColumns.tsx:
```diff
@@ -0,0 +1,64 @@
+type CellValueButtonWrapperProps = CellContext<AddIdentitiesFormData, unknown> & { refetch: () => void };
+
+function CellValueButtonWrapper(args: CellValueButtonWrapperProps) {
+    const { value: isOpen, toggle: setIsOpen } = useBoolean(false);
```
update

ml054 commented on src/Raven.Studio/typescript/test/stubs/DatabasesStubs.ts:
```diff
@@ -951,4 +951,23 @@ return docs[0];`,
+    }
+
+    private static getRandomIdentities(): Record<string, number> {
+        const randomRecord: Record<string, number> = {};
```
avoid random

M4xymm committed:
RavenDB-21764 feat (move identities to react): (CR) - moved props interface above component

RavenDB-21764 feat (move identities to react): (CR) - fixed studio conventions

RavenDB-21764 feat (move identities to react): (CR) - changed conditional to if true, rename constant

RavenDB-21764 feat (move identities to react): (CR) - changed conditional to if true

RavenDB-21764 feat (move identities to react): (CR) - moved normalizeString to generalUtils.ts

RavenDB-21764 feat (move identities to react): (CR) - moved mocks for scrollWidth and height to setup_jest.js, renamed tests to more meaningful

RavenDB-21764 feat (move identities to react): (CR) - replaced randomizing with static identities

RavenDB-21764 feat (move identities to react): removing extra pipe when we are in editing mode

RavenDB-21764 feat (move identities to react): Removed pipe from document id prefix table

RavenDB-21764 feat (move identities to react): Fixed studio convetions

RavenDB-21764 feat (move identities to react): Added equal sign to for loop to make that function more clear, simplified creating identity function to one liner.

RavenDB-21764 feat (move identities to react): changed Promise.allSettled to .All, as we handle 404 internally in configurations, in case of error throw console.error with default identity parts separator.

RavenDB-21764 feat (move identities to react): Added eventCollector, fixed pipe showing on editing

RavenDB-21764 feat (move identities to react): Studio conventions prettier fix

RavenDB-21764 feat (move identities to react): Added condition to reportEvent - call only if identity is new

RavenDB-21764 feat (move identities to react): Fix convention - added brackets to if

RavenDB-21764 feat (move identities to react): Added centered prop to Modal

ml054 commented on src/Raven.Studio/typescript/components/pages/database/documents/identities/DocumentIdentitiesModal.tsx:
```diff
@@ -0,0 +1,182 @@
+        const clientConfiguration = manageServerService.getClientConfiguration(databaseName);
+        const globalClientConfiguration = manageServerService.getGlobalClientConfiguration();
+
+        const [clientConfig, globalConfig] = await Promise.allSettled([clientConfiguration, globalClientConfiguration]);
```
I think we can use Promise.all here, as we handle 404 internally in getClient/Global configuration. 
So we either get proper value or null. 

In case of error it will simply throw.

kalczur replied:
This was my idea with allSettled, but you are right, it's actually better to handle something like that directly in command

ml054 commented on src/Raven.Studio/typescript/test/stubs/DatabasesStubs.ts:
```diff
@@ -951,4 +951,24 @@ return docs[0];`,
+
+    static getIdentities(length = 2): Record<string, number> {
+        const object: Record<string, number> = {};
+        for (let i = 1; i < length; i++) {
```
how many identities do I get when I call it with length = 5?

M4xymm replied:
due to line: 
`for (let i = 1; i < length; i++) {`

**<** means that we will have object with **4 keys**, which may be a little unclear, and I would add an **equal sign** here so that **the length equals the number of keys in the object**

ml054 commented on src/Raven.Studio/typescript/test/stubs/DatabasesStubs.ts:
```diff
@@ -951,4 +951,24 @@ return docs[0];`,
+    }
+
+    private static getStaticIdentities(index: number): Record<string, number> {
+        const staticRecord: Record<string, number> = {};
+
+        const staticKey = `key${index}`;
+        const staticValue = index * 1000;
+
+        staticRecord[staticKey] = staticValue;
+
+        return staticRecord;
```
```suggestion
        return { 
           [`key${index}`]: index * 1_000
        }
```

kalczur commented:
Add eventsCollector
https://github.com/ravendb/ravendb/blob/25ce3dca43df008b6772a2e2356f956230bbcbde/src/Raven.Studio/typescript/viewmodels/database/identities/identities.ts#L233

kalczur commented on src/Raven.Studio/typescript/components/pages/database/documents/identities/DocumentIdentitiesModal.tsx:
```diff
@@ -0,0 +1,184 @@
+    const onSubmit: SubmitHandler<AddIdentitiesFormData> = ({ prefix, value }) => {
+        return tryHandleSubmit(async () => {
+            await databasesService.seedIdentity(databaseName, prefix, value);
+            eventsCollector.reportEvent("identity", "new");
```
It should report on 'Add new identity' click. Now it is also triggered on edit identity.

M4xymm replied:
Fact, I added a condition that calls the eventCollector only when we are not in edit mode

kalczur commented on src/Raven.Studio/typescript/components/pages/database/documents/identities/DocumentIdentitiesModal.tsx:
```diff
@@ -57,7 +57,7 @@ export default function DocumentIdentitiesModal({
         return tryHandleSubmit(async () => {
             await databasesService.seedIdentity(databaseName, prefix, value);
-            eventsCollector.reportEvent("identity", "new");
+            if (!isEditing) eventsCollector.reportEvent("identity", "new");
```
brackets

kalczur replied:
I will add it to eslint rules

ml054 commented:
@M4xymm Please squash commits.

ml054 commented:
TODO: @mateuszbartosik will work on UI tweaks.

mateuszbartosik committed:
RavenDB-21764 Styling for Identities

mateuszbartosik commented:
Styling adjustments implemented
![image](https://github.com/user-attachments/assets/0c371ef7-391e-431c-bf80-1cd59c2c4c4b)

kalczur commented on src/Raven.Studio/typescript/components/pages/database/documents/identities/DocumentIdentities.stories.tsx:
```diff
@@ -28,7 +28,11 @@ export const DocumentIdentitiesStory: StoryObj<DocumentIdentitiesStoryArgs> = {
 
-        return <DocumentIdentities />;
+        return (
+            <div className="h-100">
+                <DocumentIdentities />
+            </div>
+        );
```
We should wrap all stories that way. Add this class maybe here
https://github.com/ravendb/ravendb/blob/b46a627fc2196caf732a1c53558823f8531d764a/src/Raven.Studio/.storybook/preview.js#L78

mateuszbartosik replied:
Done ✅

kalczur commented:
@M4xymm 
Don't extend `ModalProps` in `DocumentIdentitiesModalProps` interface. From what I remember, it allows you to pass any value, which makes autocomplete pointless and easier to pass the wrong name, as was the case here.

![image](https://github.com/user-attachments/assets/1d27e867-8357-43f9-aa3b-a83a743715d7)

M4xymm committed:
RavenDB-21764 feat (move identities to react): (CR) - fixed interface
